### PR TITLE
Refactoring 'children' attribute in yaml file

### PR DESCRIPTION
**Pull request by philipwinklerdelmic**

### CODE REVIEW

File: src/odemis/odemisd/modelgen.py [hunk 1]
```diff
@@ -808,7 +839,7 @@ def instantiate_component(self, name):
         # Add to the microscope all the new components that should be child
```
Comment (pieleric) [inline src/odemis/odemisd/modelgen.py]: You can update the docstring (l. 818 -> 820), to replace "children that are created by separate instantiation" by "dependencies". And drop the "if they are created by delegation".

File: src/odemis/odemisd/modelgen.py [hunk 2]
```diff
@@ -808,7 +839,7 @@ def instantiate_component(self, name):
...
         mchildren = self._microscope_ast["children"].values()
         # we only care about children created by delegation, but all is fine
-        newcmps = self.get_children(comp) # that includes comp itself
```
Comment (pieleric) [inline src/odemis/odemisd/modelgen.py]: The goal here is to catch all the newly created children. So, `get_children` should be the right thing to do (ie, don't get the dependencies).

File: src/odemis/odemisd/modelgen.py [hunk 3]
```diff
@@ -842,10 +873,11 @@ def get_instantiables(self, instantiated=None):
                 if psuname not in instantiated:
                     logging.debug("Component %s is not instantiable yet", n)
                     continue
-            for cname in attrs.get("children", {}).values():
-                childat = self.ast[cname]
+            references = attrs.get("children", {}).values() + attrs.get("dependencies", {}).values()
+            for name in references:
```
Comment (pieleric) [inline src/odemis/odemisd/modelgen.py]: IIRC, here the goal is to check whether all dependency have been instantiated (or, IOW, break if there is any dependency that hasn't been instantiated yet). So the children shouldn't matter (that's why there is the "creator" check).

File: src/odemis/odemisd/modelgen.py [hunk 4]
```diff
@@ -436,12 +442,14 @@ def _make_args(self, name):
             init["model"] = self.ast
 
         # create recursively the children
-        if "children" in init:
+        if "children" in init or "dependencies" in init:
```
Comment (pieleric) [inline src/odemis/odemisd/modelgen.py]: Just create a separate if block for "dependencies" in init, which raises a different SemanticError, with the accurate error message.

File: src/odemis/odemisd/modelgen.py [hunk 5]
```diff
@@ -453,7 +461,15 @@ def _make_args(self, name):
                 else:
                     # the child has a class or is created by another component
                     # => we explicitly reuse it
-                    init["children"][internal_role] = self._get_component_by_name(child_name)
+                    init["dependencies"][internal_role] = self._get_component_by_name(child_name)
+        if "dependencies" in init and class_name == "Microscope":
```
Comment (pieleric) [inline src/odemis/odemisd/modelgen.py]: This can never happen, as you've already checked for `"dependencies" in init`. Probably you wanted to check for `"dependencies" in attr`?

File: src/odemis/odemisd/modelgen.py [hunk 6]
```diff
@@ -453,7 +461,15 @@ def _make_args(self, name):
...
+            raise SemanticError("Error in microscope file: "
+                                "microscope class '%s' should not have 'dependencies' argument." % name)
+        if "dependencies" in attr:
+            init["dependencies"] = {}
```
Comment (pieleric) [inline src/odemis/odemisd/modelgen.py]: I'm a little confused how the legacy depedencies are handled. IMHO, it'd be best if children which have a class (or a creator which is not the given component) are automatically moved to the "dependencies". However, from the code just above, it seems you're just handling them on-the-fly the old way. It should work too. However, if there is a mix legacy + new dependencies, the first init["dependencies"] (from the legacy) is going to be silently dropped. I'd suggest to either raise an Error, or merge them, and log a warning.

File: src/odemis/odemisd/main.py [hunk 7]
```diff
@@ -288,7 +288,7 @@ def _instantiate_component(self, name):
                 pass
             raise ValueError("Failed to instantiate component %s" % name)
         else:
-            children = self._instantiator.get_children(comp)
```
Comment (pieleric) [inline src/odemis/odemisd/main.py]: Technically, here, we only care about the children anyway. So we don't need to get also the dependencies, and then filter them with the "delegated_children". => just call `get_children()` and add them to the alive.

BTW, you could probably rename `get_delegated_children()` to something without "delegated"... maybe `get_children_names()`.

File: src/odemis/model/_core.py [hunk 8]
```diff
@@ -128,6 +128,8 @@ def _getChildren(root):
     ret = {root}
     for child in root.children.value:
         ret |= _getChildren(child)
+    for dep in root.dependences.value:
```
Comment (pieleric) [inline src/odemis/model/_core.py]: I don't think this is needed. The point is to get all the components created (and which are useful). So the (delegated) children should be sufficient.

File: src/odemis/model/_components.py [hunk 9]
```diff
@@ -107,12 +107,16 @@ class Component(ComponentBase):
     Component to be shared remotely
     '''
 
-    def __init__(self, name, parent=None, children=None, daemon=None):
+    def __init__(self, name, parent=None, children=None, dependencies=None, daemon=None):
         """
         name (string): unique name used to identify the component
         parent (Component): the parent of this component, that will be in .parent
         children (dict str -> Component): the children of this component, that will
-            be in .children. Objects not instance of Component are skipped
```
Comment (pieleric) [inline src/odemis/model/_components.py]: I don't think this is correct. The idea of the "children" argument is that if the component has already created the children, it can directly pass them, and the VA will be initialised with them already.
For the microscope class, it works as a weird hack, but let's not mention it here. Only in the Microscope class.

File: src/odemis/model/_components.py [hunk 10]
```diff
@@ -107,12 +107,16 @@ class Component(ComponentBase):
...
+            be in .children. Objects not instance of Component are skipped. This means that
+            this argument is only relevant for the microscope class since all other classes
+            don't have component children (they are dependencies).
+        dependencies (dict str -> Component): the dependencies of this component, that will
```
Comment (pieleric) [inline src/odemis/model/_components.py]: The `dependencies` argument should only get Components. So I'd suggest to raise an Error is the objects are not Component (instead of silently skipping them).

File: src/odemis/driver/simcam.py [hunk 11]
```diff
@@ -58,7 +58,7 @@ def __init__(self, name, role, image, children=None, daemon=None, **kwargs):
         converter = dataio.find_fittest_converter(image, mode=os.O_RDONLY)
         self._img = converter.read_data(image)[0]  # can be RGB or greyscale
 
-        # we will fill the set of children with Components later in ._children
+        # we will fill the set of dependencies with Components later in ._dependencies
```
Comment (pieleric) [inline src/odemis/driver/simcam.py]: No, from the code, it doesn't seem we fill the set of dependencies afterwards. You need to pass dependencies too here.

File: src/odemis/driver/scanner.py [hunk 12]
```diff
@@ -26,39 +26,39 @@
...
+        extnl = dependencies["external"]
         if not isinstance(extnl, ComponentBase):
-            raise ValueError("Child external is not a component.")
+            raise ValueError("Dependency external is not a component.")
         if not model.hasVA(extnl, "pixelSize"):
-            raise ValueError("Child external is not a Emitter component.")
+            raise ValueError("Dependency external is not a Emitter component.")
         self._external = extnl
-        self.children.value.add(extnl)
+        self.dependencies.value.add(extnl)
```
Comment (pieleric) [inline src/odemis/driver/scanner.py]: Just pass "dependencies" to the super class, and you don't need to update it here.

File: src/odemis/driver/picoquant.py [hunk 13]
```diff
@@ -209,18 +209,21 @@ class PH300(model.Detector):
...
-    def __init__(self, name, role, device=None, children=None, daemon=None,
+    def __init__(self, name, role, device=None, dependencies=None, children=None, daemon=None,
                  disc_volt=None, zero_cross=None, shutter_axes=None, **kwargs):
         """
         device (None or str): serial number (eg, 1020345) of the device to use
           or None if any device is fine.
-        children (dict str -> kwargs): the names of the detectors (detector0 and
-         detector1 are valid) to the arguments.
+        dependencies (dict str -> kwargs): shutters components (shutter0 and shutter1 are valid)
+        children (dict str -> dict): the names of the detectors (detector0 and
```
Comment (pieleric) [inline src/odemis/driver/picoquant.py]: dict str -> kwargs.

File: src/odemis/driver/picoquant.py [hunk 14]
```diff
@@ -209,18 +209,21 @@ class PH300(model.Detector):
...
 
-    def __init__(self, name, role, device=None, children=None, daemon=None,
+    def __init__(self, name, role, device=None, dependencies=None, children=None, daemon=None,
                  disc_volt=None, zero_cross=None, shutter_axes=None, **kwargs):
         """
         device (None or str): serial number (eg, 1020345) of the device to use
           or None if any device is fine.
-        children (dict str -> kwargs): the names of the detectors (detector0 and
-         detector1 are valid) to the arguments.
+        dependencies (dict str -> kwargs): shutters components (shutter0 and shutter1 are valid)
```
Comment (pieleric) [inline src/odemis/driver/picoquant.py]: dict str -> Component

File: src/odemis/driver/picoquant.py [hunk 15]
```diff
@@ -268,20 +271,23 @@ def __init__(self, name, role, device=None, children=None, daemon=None,
...
+                if "shutter1" in dependencies:
                     shutter_name = "shutter1"
                 else:
                     shutter_name = None
                 self._detectors[name] = PH300RawDetector(channel=1, parent=self, shutter_name=shutter_name, daemon=daemon, **ckwargs)
                 self.children.value.add(self._detectors[name])
-            elif name == "shutter0":
+            else:
+                raise ValueError("Child %s not recognized, should be detector0 or detector1.")
+        for name, ckwargs in dependencies.items():
```
Comment (pieleric) [inline src/odemis/driver/picoquant.py]: Rename ckwargs -> comp

File: src/odemis/driver/phenom.py [hunk 16]
```diff
@@ -202,63 +203,63 @@ def __init__(self, name, role, children, host, username, password, phenom_gui=Fa
         except ValueError:
             logging.warning("Phenom version could not be retrieved")
 
-        # Lock in order to synchronize all the child component functions
```
Comment (pieleric) [inline src/odemis/driver/phenom.py]: Every child in this class is really a delegated child, as you noted by not modifying the code. => Revert the changes to the comments and messages.

File: src/odemis/driver/hamamatsurx.py [hunk 17]
```diff
@@ -877,7 +877,7 @@ class StreakCamera(model.HwComponent):
     Client to connect to HPD-TA software via RemoteEx.
     """
 
-    def __init__(self, name, role, port, host, children=None, daemon=None, **kwargs):
```
Comment (pieleric) [inline src/odemis/driver/hamamatsurx.py]: All the changes done on hamamatsurx should be done identically to simstreakcam.py.

File: src/odemis/driver/hamamatsurx.py [hunk 18]
```diff
@@ -877,7 +877,7 @@ class StreakCamera(model.HwComponent):
...
+    def __init__(self, name, role, port, host, dependencies=None, daemon=None, **kwargs):
```
Comment (pieleric) [inline src/odemis/driver/hamamatsurx.py]: As you've properly done in the YAML file for the streakcam, it needs both children and dependencies. Almost all the components are children, and the only one special is the "spectrograph", which is an (optional) dependency.

File: src/odemis/odemisd/modelgen.py (hunk 3 above)
Comment (philipwinklerdelmic) [reply src/odemis/odemisd/modelgen.py]: I agree. We still need it for backwards compatibility with the old children though, right?

**Comment by philipwinklerdelmic** [comment]
pushed new changes

**Comment by pieleric** [comment]
I've just found another place that needs to be updated: model._core.ContainerObject.instantiate(): change all the "children" to "dependencies".

**Comment by philipwinklerdelmic** [comment]
done

File: src/odemis/model/_components.py [hunk 19]
```diff
@@ -107,12 +107,14 @@ class Component(ComponentBase):
...
-    def __init__(self, name, parent=None, children=None, daemon=None):
+    def __init__(self, name, parent=None, children=None, dependencies=None, daemon=None):
         """
         name (string): unique name used to identify the component
         parent (Component): the parent of this component, that will be in .parent
         children (dict str -> Component): the children of this component, that will
-            be in .children. Objects not instance of Component are skipped
+            be in .children. Objects not instance of Component are skipped.
+        dependencies (dict str -> Component): the dependencies of this component, that will
+            be in .dependencies. Objects not instance of Component are skipped
```
Comment (pieleric) [inline src/odemis/model/_components.py]: Please drop this "Objects not instance of Component are skipped", as it's not correct.

File: src/odemis/model/_core.py [hunk 20]
```diff
@@ -234,19 +234,19 @@ def instantiate(self, klass, kwargs):
         returns the new component instantiated
         """
         try:
-            # HACK: we know that children can contain components, which are pretty
+            # HACK: we know that dependencies can contain components, which are pretty
```
Comment (pieleric) [inline src/odemis/model/_core.py]: "can contain" -> "should contain".

File: src/odemis/odemisd/modelgen.py [hunk 21]
```diff
@@ -453,7 +464,18 @@ def _make_args(self, name):
...
-                    init["children"][internal_role] = self._get_component_by_name(child_name)
+                    init["dependencies"][internal_role] = self._get_component_by_name(child_name)
+        if "dependencies" in attr and class_name == "Microscope":
+            raise SemanticError("Error in microscope file: "
+                                "microscope class '%s' should not have 'dependencies' argument." % name)
+        if "dependencies" in attr:
+            if "dependencies" not in init:
+                # in case of a mixed yaml-file with old-style and new-style children, the 'dependencies'
+                # argument might already exist.
+                init["dependencies"] = {}
```
Comment (pieleric) [inline src/odemis/odemisd/modelgen.py]: Add a warning if there are both legacy and new style mixed. Something like:
```
else:
   logging.warning("Mix legacy dependent-component (%s) + dependencies", init["dependencies"].keys)
```

File: src/odemis/odemisd/modelgen.py [hunk 22]
```diff
@@ -842,10 +878,11 @@ def get_instantiables(self, instantiated=None):
                 if psuname not in instantiated:
                     logging.debug("Component %s is not instantiable yet", n)
                     continue
-            for cname in attrs.get("children", {}).values():
-                childat = self.ast[cname]
+            references = attrs.get("children", {}).values() + attrs.get("dependencies", {}).values()
```
Comment (pieleric) [inline src/odemis/odemisd/modelgen.py]: Please add a comment explaining that in theory, we only care about the dependencies, but to handle legacy code which have their dependencies as "children", we also include children, and check if the creator is not the component.
Alternatively, you could write more explicitly:
```
dep = list(attrs.get("dependencies", {}).values())
deps += [c for c in attrs.get("children", {}).values() if self.ast[c].get("creator") != n]
for name in deps:
    if name not in instantiated:
        break
```
It'd make the code closer from the "final" code when the legacy support is removed.

File: src/odemis/odemisd/modelgen.py [hunk 23]
```diff
@@ -600,12 +631,17 @@ def get_required_components(self, name):
...
+            ret.add(n)
+            ret |= self.get_required_components(n)
+
         try:
-            children.append(attrs["power_supplier"])
+            dependencies.append(attrs["power_supplier"])
         except KeyError:
             pass  # no power supplier
 
+        children = attrs.get("children", {}).values()
```
Comment (pieleric) [inline src/odemis/odemisd/modelgen.py]: Add a comment that this is to support legacy code.

**Comment by philipwinklerdelmic** [comment]
pushed fixes

File: src/odemis/model/_components.py (hunk 19 above)
Comment (pieleric) [reply src/odemis/model/_components.py]: Still there?

Comment (pieleric) [reply src/odemis/model/_components.py]: Oh, actually, you've removed it... but the wrong one. Please, put back the text for children, and remove for dependencies.

Comment (philipwinklerdelmic) [reply src/odemis/model/_components.py]: Indeed, I didn't pay attention... Should be fixed now.
